### PR TITLE
remove redundant medoids number k in Python fit method. Fixes #48

### LINKED ISSUE #48: Examples in README pass `k` to `kmed.fit` even though `KMedoids` object already has this parameter
image

In the image above, there is redundancy with the number of medoids `3`. This can create confusion when there are discrepancies. Please remove the redundancy to make the `README` clearer

### PR DESCRIPTION
**Pull request by mailology**
The argument `k` in Python fit method is redundant since we have already specified the medoids number when we instantiate the `KMedoids` class. We therefore remove this argument in `fitPython()` method in `KMedsWrapper` class. The Python examples in repo are tested again using the updated code and the results are as follow:

BanditPAM_GMM_new_code

BanditPAM_MNIST_new_code

The standard cases `./BanditPAM -f ../../data/MNIST-1k.csv -k 10 -v 1` with different loss functions `manhattan, cos, inf, L15` are also tested and they all give the expected results.

### CODE REVIEW

File: src/kmeds_pywrapper.cpp [hunk 1]
```diff
@@ -35,11 +35,9 @@ class KMedsWrapper : public KMedoids {
    *
    * @param inputData Input data to find the medoids of
    * @param loss The loss function used during medoid computation
-   * @param k The number of medoids to compute
    * @param logFilename The name of the outputted log file
    */
-  void fitPython(py::array_t<double> inputData, std::string loss, int k, std::string logFilename) {
-    KMedoids::setNMedoids(k);
+  void fitPython(py::array_t<double> inputData, std::string loss, std::string logFilename) {
```
Comment (motiwari) [inline src/kmeds_pywrapper.cpp:40]: Ah, I think this is the wrong thing to do. We *want* the user to be able to call `.fit()` from Python with a specific number of medoids, or instantiate the `KMedoids` object with a value for `k`. 

#48  was actually just referring to changing the `README` documentation -- not the C++ or Python code, just the example that uses `k=3` in two places instead of one

**Review by motiwari — CHANGES_REQUESTED** [review]

File: src/kmeds_pywrapper.cpp (hunk 1 above)
Comment (mailology) [reply src/kmeds_pywrapper.cpp]: Oh, it make senses, I think I have overthought the issue!

File: README.md [hunk 2]
```diff
@@ -54,9 +54,10 @@ means = np.array([[0,0], [-5,5], [5,5]])
 X = np.vstack([np.random.randn(n_per_cluster, 2) + mu for mu in means])
 
 # Fit the data with BanditPAM:
-kmed = KMedoids(n_medoids = 3, algorithm = "BanditPAM")
+k = 3
+kmed = KMedoids(n_medoids = k, algorithm = "BanditPAM")
 # Writes results to gmm_log
-kmed.fit(X, 'L2', 3, "gmm_log")
+kmed.fit(X, 'L2', k, "gmm_log")
```
Comment (motiwari) [inline README.md:60]: The point of this issue is that passing the argument `k` here is redundant -- it is already passed in line 58. So, we can remove the passing of `k` in line 60.

File: README.md [hunk 3]
```diff
@@ -87,8 +88,9 @@ X = pd.read_csv('data/MNIST-1k.csv', sep=' ', header=None).to_numpy()
 X_tsne = TSNE(n_components = 2).fit_transform(X)
 
 # Fit the data with BanditPAM:
-kmed = KMedoids(n_medoids = 10, algorithm = "BanditPAM")
-kmed.fit(X, 'L2', 10, "mnist_log")
+k = 10
+kmed = KMedoids(n_medoids = k, algorithm = "BanditPAM")
+kmed.fit(X, 'L2', k, "mnist_log")
```
Comment (motiwari) [inline README.md:93]: Same here; passing the argument `k` is redundant because it is already passed in Line 92

**Review by motiwari — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (mailology) [reply README.md:60]: Hmmm, since the `fit` function requires four arguments, I may need to modify the python code then?

Comment (motiwari) [reply README.md:60]: We should neither require the parameter when instantiating the `KMedoids` object or when calling `.fit()`. Only if we call `.fit()` without the parameter, and the parameter has not already been set, should we raise an error. Could you modify the code to reflect this?

**Comment by mailology** [comment]
The `fitPython()` function is updated so that the number of medoids `k` becomes optional. If `n_medoids` is specified when instantiating the `KMedoids` object, this value is used implicitly in the `.fit()` function. However, if `k` is specified in `.fit()`, then this value will be used as the number of medoids and overwrite any previous assigned `n_medoids`.  

Here are the examples I have tried:
(i) `n_medoids` is specified in `KMedoids` but `k` is not specified
Screen Shot 2021-06-26 at 11 46 33 PM

(ii) `n_medoids` is specified in `KMedoids` and `k` is also specified but of different value (overwrite `n_medoids`)
Screen Shot 2021-06-26 at 11 50 46 PM

(iii) `n_medoids` is not specified but `k` is specified 
Screen Shot 2021-06-26 at 11 52 29 PM

(iv) Neither `n_medoids` nor `k` is specified 
Screen Shot 2021-06-26 at 11 54 16 PM

File: src/kmedoids_ucb.cpp [hunk 4]
```diff
@@ -35,6 +35,15 @@ KMedoids::KMedoids(int n_medoids, std::string algorithm, int verbosity,
   KMedoids::checkAlgorithm(algorithm);
 }
 
+KMedoids::KMedoids(std::string algorithm, int verbosity,
```
Comment (motiwari) [inline src/kmedoids_ucb.cpp:38]: We shouldn't duplicate code: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

Instead, could you pass a default argument of null to the original function? Then, inside the function surface an error if that argument is still null.

File: src/kmeds_pywrapper.cpp [hunk 5]
```diff
@@ -38,8 +38,25 @@ class KMedsWrapper : public KMedoids {
...
    * @param logFilename The name of the outputted log file
    */
-  void fitPython(py::array_t<double> inputData, std::string loss, int k, std::string logFilename) {
-    KMedoids::setNMedoids(k);
+  void fitPython(py::array_t<double> inputData, std::string loss, std::string logFilename, py::kwargs kw) {
+    // throw an error if the number of medoids is not specified in either the KMedoids object or the fitPython function
+    try {
+      if (!KMedoids::getNMedoids()) {
+        if (kw.size() == 0) {
+          throw py::value_error("n_medoids");
```
Comment (motiwari) [inline src/kmeds_pywrapper.cpp:46]: Could we make the error message more informative to the user?

File: src/kmeds_pywrapper.cpp [hunk 6]
```diff
@@ -38,8 +38,25 @@ class KMedsWrapper : public KMedoids {
...
+        }
+      }
+    } catch (py::error_already_set &c) {
+       // py::value_error is not a Python exception
+      assert(false);
```
Comment (motiwari) [inline src/kmeds_pywrapper.cpp:51]: What does this do?

**Review by motiwari — CHANGES_REQUESTED** [review]
Thank you for the updates, @mailology ! We're getting there -- see comments in the review.

File: src/kmeds_pywrapper.cpp (hunk 6 above)
Comment (mailology) [reply src/kmeds_pywrapper.cpp]: The error catch follows from an example in pybind11 exceptions document: https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#handling-exceptions-from-python-in-c

When C++ calls Python functions, such as in a callback function or when manipulating Python objects, and Python raises an Exception, pybind11 converts the Python exception into a C++ exception of type `pybind11::error_already_set ` whose payload contains a C++ string textual summary and the actual Python exception. `error_already_set` is used to propagate Python exception back to Python (or possibly, handle them in C++).

I have removed this part and the error message is still raised properly. It seems we don't need this part.

**Comment by mailology** [comment]
I have passed a default argument for `n_medoids` as null and update the logic in `.fitPython()` so that it raises an error if the argument is still null. The corresponding constructors for the repeated part are removed. The error message is made more informative, and the error catch for `error_already_set` is redundant so it is removed.  

The four different Python cases are checked again and the results are the same except the last one has error message improved: 
Screen Shot 2021-06-28 at 11 05 26 PM

File: README.md [hunk 7]
```diff
@@ -87,8 +88,9 @@ X = pd.read_csv('data/MNIST-1k.csv', sep=' ', header=None).to_numpy()
 X_tsne = TSNE(n_components = 2).fit_transform(X)
 
 # Fit the data with BanditPAM:
-kmed = KMedoids(n_medoids = 10, algorithm = "BanditPAM")
-kmed.fit(X, 'L2', 10, "mnist_log")
+k = 10
```
Comment (motiwari) [inline README.md:91]: ordinarily it's good practice to define variables and avoid magic numbers, but in this case for the README examples, since `k` is only used once, let's not use a separate variable and just pass `10` in the call below

File: README.md [hunk 8]
```diff
@@ -54,9 +54,10 @@ means = np.array([[0,0], [-5,5], [5,5]])
 X = np.vstack([np.random.randn(n_per_cluster, 2) + mu for mu in means])
 
 # Fit the data with BanditPAM:
-kmed = KMedoids(n_medoids = 3, algorithm = "BanditPAM")
+k = 3
```
Comment (motiwari) [inline README.md:57]: ordinarily it's good practice to define variables and avoid magic numbers, but in this case for the README examples, since `k` is only used once, let's not use a separate variable and just pass `3` in the call below

**Review by motiwari — CHANGES_REQUESTED** [review]
Looks great! A few nits -- can you make the changes and rebase on top of `main`?

**Comment by mailology** [comment]
Fixed the nits on readme. I have also done the rebase on top of `main`.

File: src/main.cpp [hunk 9]
```diff
@@ -45,7 +46,7 @@ int main(int argc, char* argv[])
                 f_flag = true;
                 break;
             // path to log output data
```
Comment (motiwari) [inline src/main.cpp:48]: update comments

File: src/kmeds_pywrapper.cpp [hunk 10]
```diff
@@ -38,8 +38,22 @@ class KMedsWrapper : public KMedoids {
    * @param k The number of medoids to compute
    * @param logFilename The name of the outputted log file
    */
-  void fitPython(py::array_t<double> inputData, std::string loss, int k, std::string logFilename) {
-    KMedoids::setNMedoids(k);
+  void fitPython(py::array_t<double> inputData, std::string loss, std::string logFilename, py::kwargs kw) {
+    // throw an error if the number of medoids is not specified in either the KMedoids object or the fitPython function
```
Comment (motiwari) [inline src/kmeds_pywrapper.cpp:42]: break onto separate line

**Review by motiwari — CHANGES_REQUESTED** [review]
Overall looking much better! 2 nits about the comments and then should be good to merge

**Comment by mailology** [comment]
Fixes some nits on comments.

**Review by motiwari — APPROVED** [review]
